### ISSUE TITLE
Rate limiting when scheduling proxy client requests

### ISSUE THREAD

**Issue by stevenviola** [sensu/sensu#1650]
## Context

If a proxy client request is configured to check an external resource, such as an API for each client, this can overwhelm the API if there are a lot of clients which match, and create many checks to be made and executed in a short amount of time. This creates a thundering heard, and can result in API timeouts or exceeding API rate limiting.

Changing the interval doesn't mitigate the problem, and just creates bursts of jobs at the set interval. A solution which would distribute jobs over time would reduce the bursty nature of scheduling the checks

Currently, I'm experiencing this with a large number of clients in AWS, and am using a proxy check for checking the AWS API for specific resources per client. This results in many requests to happen all at once against the API, and leads to being blocked by the AWS API, and some checks to randomly fail.

## Expected Behavior

Optional check attribute is added which limits the rate that check requests are processed simultaneously. Setting this value to 10 for example would ensure that the check is only scheduled to be executed a maximum of 10 checks per second. Omitting this value would result in no rate limiting.

## Current Behavior

With no rate limiting in place, check requests which apply to many clients can potentially DDoS external services. 

## Possible Solution

If clients honor the issued timestamp in the check request, and nack the check if the issued time is in the future, then on the server end, adding seconds (based on the rate limiting attribute of the check) to the issued timestamp will result in issued times to be in the future while quickly publishing checks to the transport, and not require sleeping in the loop to wait for the right time to publish the proxy check.

## Your Environment
* Sensu version used: **0.28.5**
* Operating System and version: **CentOS 6.8**

**Comment by portertech** [comment]
@stevenviola are you not able to split the proxy check up into AWS resource specific checks? This would alleviate the problem.

**Comment by stevenviola** [comment]
I am using the sensu-plugins-aws plugin, and running check-instance-health.rb for each instance. This creates an API request per instance. I need to have the check results have the source of the instance for notification purposes, and to also provide meaningful content. 

My proxy check request looks like the following:

```
{
  "checks": {
    "check-instance-health": {
      "command": "check-instance-health.rb -f '{name:instance-id,values:[:::ec2-id:::]}'",
      "handlers": [
        "mailer"
      ],
      "interval": 600,
      "occurrences": 5,
      "refresh": 7200,
      "subscribers": [
        "roundrobin:aws-checks"
      ],
      "standalone": false,
      "ttl": 1800,
      "proxy_requests": {
        "client_attributes": {
          "instance-type": "ec2"
        }
      }
    }
  }
}
```

Defining a check per instance would be way too much to maintain in a dynamic environment, and with so many hosts. Even if this were not a proxy check request, when the sensu server schedules the check, all the hosts will check themselves against the API at the same time, and basically have the same impact. I figured with the proxy check request, using a round robin would add some delay, but it was not enough to avoid the AWS rate limiting.

**Comment by runningman84** [comment]
I really ask myself if running these aws cloudwatch based checks really makes sense. It might be better to use cloudwatch alarms and send the results to an sns topic which calls a lambda function and creates a sensu event? This is a general concern because the aws checks consume a lot of cpu resources ...

**Comment by russellshackleford** [comment]
@runningman84, while I agree with your concerns, on a large scale dynamic environment, it is very useful to have all of your alerts going through the same portal for the operations teams. Additionally, using sensu to do these checks allows for a more centralized configuration management which is also very useful when you have many disconnected teams all trying to achieve the same results.

**Comment by stevenviola** [comment]
@runningman84 yeah, I have noticed higher CPU usage when using the aws sdk. This is off topic, but do you suggest having the lambda function parse the Sensu API for the clients which match, get the stats from cloudwatch, and then use the /results endpoint to publish the results? I can see this as a viable solution, so I'll investigate to get around the current issue I'm having.

As for the rate limiting, proxy request checks introduce the ability to create many individual checks, which are all executed on the clients at the same time, which is the crux of the concern. For example, in the SNMP pollers example from the docs, if you have a lot of snmp clients in your infrastructure which need to be checked, the pollers who execute the checks will issue many snmp requests at once, which can create lots of traffic and essentially create an SNMP Amplification attack within the local network. Some method of spacing out the requests would decrease the burstiness of the checks, regardless of what the check is trying to run.

Hope that provides some additional context to this.

**Comment by runningman84** [comment]
@stevenviola yes I think we should create a lambda function which will run hourly or is attached to aws config for service discovery. This lambda function should create corresponding "fake" clients in sensu for elb, rds, dynamodb and other AWS services. Furthermore it should create cloudwatch alarms which would call a lambda function which would create sensu events.

This kind of solution would reduce the cpu usage on sensu servers a lot...

**Comment by portertech** [comment]
I believe we can implement a simple opt-in feature that gets us something that resembles rate limiting, providing part of the functionality. Proxy requests "splay" (boolean) would tell Sensu to distribute proxy check requests within a time window determined by the check "interval" and the number of clients in the registry that match, e.g. `"proxy_requests": {"splay": true}` -> `splay = check_interval * .9 / client_count`.

**Comment by mcbsdEU** [comment]
This would be really helpful to have.

**Comment by portertech** [comment]
Closed by #1756 🎉 

![best-thing-ever-hand](https://user-images.githubusercontent.com/149630/30932044-3025ecce-a37b-11e7-9ebd-d448f684fb3e.gif)
